### PR TITLE
avro serialization implement

### PR DESCRIPTION
**Pull request by playaround88**
hi,  
I saw same todo task in docs, so I implements the avro serializtion.
I have copy and change some simple junit test.   
and I have used dubbo-demo test some param case like Person Object, String Array. All is right work!

if be merged, it's my honor!   
I'll try hard to contribute more.   
Thx

## What is the purpose of the change
avro serialization implement

## Brief changelog
* dubbo-serialization/pom.xml
* three components implement of the api
* test unit

## Verifying this change

yes

Follow this checklist to help us incorporate your contribution quickly and easily:

- [ ] Make sure there is a [GITHUB_issue](https://github.com/apache/incubator-dubbo/issues) filed for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [ ] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add integration-test in [test module](https://github.com/apache/incubator-dubbo/tree/master/dubbo-test).
- [x] Run `mvn clean install -DskipTests` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/incubator-dubbo/wiki/Software-donation-guide).

### CODE REVIEW

File: dubbo-serialization/dubbo-serialization-avro/pom.xml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+	<dependencies>
+		<dependency>
+			<groupId>org.apache.dubbo</groupId>
+			<artifactId>dubbo-serialization-api</artifactId>
+			<version>${project.parent.version}</version>
+		</dependency>
+		<dependency>
+			<groupId>org.apache.avro</groupId>
+			<artifactId>avro</artifactId>
+			<version>1.8.2</version>
```
Comment (kimmking) [inline dubbo-serialization/dubbo-serialization-avro/pom.xml]: move version to dependency-bom

File: dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectInput.java [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+public class AvroObjectInput implements ObjectInput{
+	private DecoderFactory decoderFactory=DecoderFactory.get();;
+	private BinaryDecoder decoder;
+	
+	public AvroObjectInput(InputStream in){
+		init(in);
+	}
+	
+	public void init(InputStream in){
```
Comment (kimmking) [inline dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectInput.java]: move to method AvroObjectInput(InputStream in)

File: dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectOutput.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+	private EncoderFactory encoderFactory=EncoderFactory.get();
+	private BinaryEncoder encoder;
+	
+
+    public AvroObjectOutput(OutputStream out) {
+    	this.out=out;
+    	init(out);
+    }
+
+    public void init(OutputStream out){
```
Comment (kimmking) [inline dubbo-serialization/dubbo-serialization-avro/src/main/java/org/apache/dubbo/common/serialize/avro/AvroObjectOutput.java]: move init to constructor

File: dubbo-serialization/dubbo-serialization-avro/src/test/java/org/apache/dubbo/common/serialize/avro/AvroObjectOutputTest.java [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package org.apache.dubbo.common.serialize.avro;
+
+public class AvroObjectOutputTest {
```
Comment (kimmking) [inline dubbo-serialization/dubbo-serialization-avro/src/test/java/org/apache/dubbo/common/serialize/avro/AvroObjectOutputTest.java]: no used？

**Comment by playaround88** [comment]
hi kimmking,
I have improved code for you review, and pushed a new commit. 
Sorry and Thx!

detail: 
* add avro-serialization to bom/pom.xml
* move avro version to dependencies-bom/pom.xml
* remove init method in AvroObjectInput/AvroObjectOutput
* remove no use file in test. AvroObjectOutputTest.java, model/Image.java, users.avro
* add @SuppressWarnings for AvroObjectInput.readObject and AvroObjectOutput.writeObject

**Comment by playaround88** [comment]
@diecui1202 
Sorry, I do not understand your meaning.
1. dubbo now is under apache/incubator-dubbo, and I fork from here.
2. this RP is as you say commit as a seperate project named dubbo-serialization-avro.
3. I can find dubbo project at github.com/dubbo, which project should I contribute on?

first time to post RP, and I have found some bad ops on PR(this RP is on my master branch).

thanks and please express！

**Comment by diecui1202** [comment]
@playaround88 

You can check this repo: https://github.com/dubbo/dubbo-serialization-native-hessian

Now we recommend the extensions of dubbo are placed in dubbo ecosystem, so for your avro extension, you don't need to work on any existing project, just create a new project named dubbo-serialization-avro in your own repository. Then we transfer it to ecosystem.

**Comment by playaround88** [comment]
@diecui1202 
Thank you, I got it.
I will do that and after paste project here.

**Comment by playaround88** [comment]
@diecui1202 @chickenlj @kimmking 
hi all,
I have move the RP to an seperate project. and I will resend to mail-list to notify the community.
addressed here: https://github.com/playaround88/dubbo-serialization-avro

**Comment by diecui1202** [comment]
cool~

**Pull request closed without merging**